### ISSUE TITLE
Trigger tech event when master playlist failed to load

### ISSUE THREAD

**Issue by arski** [videojs/videojs-contrib-hls#1332]
Hi there, I'm wondering if we could add a line like 

    this.tech_.trigger('nomasterplaylist');

somewhere around https://github.com/videojs/videojs-contrib-hls/blob/master/src/master-playlist-controller.js#L864

Because this is a pretty frequent case with broken CORS headers, but all I can rely on is the tech.trigger('error', hlserror), so I have to set a handler on the very generic 'error' event and then filter out by "HLS" being in the error message which is not.. ideal..

What do you think?

cc @robinmaben

**Comment by squarebracket** [comment]
Could you elaborate on the problem that you're trying to catch? I'm not sure I understand the context.

**Comment by arski** [comment]
Sure! So basically we're trying to download the master playlist and the request fails because of broken CORS headers (I believe this is a known issue for some CDNs on Safari/Mac, Edge and Chrome/Android).

So basically we end up here https://github.com/videojs/videojs-contrib-hls/blob/master/src/playlist-loader.js#L231

which triggers https://github.com/videojs/videojs-contrib-hls/blob/master/src/master-playlist-controller.js#L445

However the blacklist function ends up in https://github.com/videojs/videojs-contrib-hls/blob/master/src/master-playlist-controller.js#L862 and closes up.. and never reaches e.g. this.tech_.trigger('blacklistplaylist');

So bottom line is, if the request for the master playlist fails, there is no unique tech event being fired, making it very hard to identify from the outside and e.g. log somewhere, or handle somehow.

**Comment by squarebracket** [comment]
Hmm. Well, firing an event for failing to load the master playlist isn't totally unreasonable.

But, it sounds to me like the problem is on the first load, so couldn't you do something like this?
```
var initialError = () => { console.error("Failed to load media");
player.on('error', initialError);
player.one('loadedmetadata', () => { player.off('error', initialError); });
```

Alternatively, you could do something like an `OPTIONS` request to the server to make sure the CORS config is valid before passing the master playlist URL on to video.js.

Any chance you have a source for that known CORS issue? Would be good to know about it, but since CORS is so finicky to set up I have a feeling googling will just lead me to a zillion stack overflows about how to set up CORS.

**Comment by arski** [comment]
Hi, yea our team is working on fixing the CORS headers but I just want to be able to dynamically catch the problem whenever it pops up, e.g. if something breaks again. I think it's generally related to issues like https://github.com/videojs/videojs-contrib-hls/issues/1162#issuecomment-315517802

What you're suggesting would work I guess, but it feels like a conditional workaround, just because we know it's on first load we can rely on the generic 'error'. I imagine there are other errors that can pop up during the player initialization up until the loadedmetadata point, so I'm a bit hesitant about classifying them all as "HLS master playlist failed" if you see what I mean.

**Comment by gesinger** [comment]
@squarebracket 's workaround is a good one, but it seems reasonable to want to a more precise error (for instance, that initial error could be caused by a different issue).

We should add either an event or a more description error in this case.
